cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i

```diff
@@ -0,0 +1,214 @@
+    DAAL_INT liwork = 5 * nFeatures + 3;
+    DAAL_INT info;
+
+    TArray<algorithmFPType, cpu> work(lwork);
+    TArray<DAAL_INT, cpu> iwork(liwork);
+    DAAL_CHECK_MALLOC(work.get() && iwork.get());
+
+    LapackInst<algorithmFPType, cpu>::xsyevd(&jobz, &uplo, (DAAL_INT *)(&nFeatures), eigenvectors, (DAAL_INT *)(&nFeatures), eigenvalues, work.get(),
+                                             &lwork, iwork.get(), &liwork, &info);
+    if (info != 0) return services::Status(services::ErrorPCAFailedToComputeCorrelationEigenvalues); // CHANGE ERROR STATUS
+    return services::Status();
+}
+
+/**
+ *  \brief Kernel for Spectral Embedding calculation
+ */
+template <typename algorithmFPType, Method method, CpuType cpu>
+services::Status SpectralEmbeddingKernel<algorithmFPType, method, cpu>::compute(const NumericTable * xTable, NumericTable * embeddingTable,
+                                                                                NumericTable * eigenTable, const KernelParameter & par)
+{
+    services::Status status;
+    // std::cout << "inside DAAL kernel" << std::endl;
+    // std::cout << "Params: " << par.numberOfEmbeddings << " " << par.numberOfNeighbors << std::endl;
+    size_t k       = par.numberOfEmbeddings;
+    size_t filtNum = par.numberOfNeighbors + 1;
+    size_t n       = xTable->getNumberOfRows(); /* Number of input feature vectors   */
+
+    SharedPtr<HomogenNumericTable<algorithmFPType> > tmpMatrixPtr =
+        HomogenNumericTable<algorithmFPType>::create(n, n, NumericTable::doAllocate, &status);
+
+    DAAL_CHECK_STATUS_VAR(status);
+    NumericTable * covOutput   = tmpMatrixPtr.get();
+    NumericTable * a0          = const_cast<NumericTable *>(xTable);
+    NumericTable * eigenvalues = const_cast<NumericTable *>(eigenTable);
+
+    // Compute cosine distances matrix
+    {
+        auto cosDistanceKernel = cosine_distance::internal::DistanceKernel<algorithmFPType, cosine_distance::Method::defaultDense, cpu>();
+        DAAL_CHECK_STATUS(status, cosDistanceKernel.compute(0, &a0, 0, &covOutput, nullptr));
+    }
+
+    WriteRows<algorithmFPType, cpu> xMatrix(covOutput, 0, n);
+    DAAL_CHECK_BLOCK_STATUS(xMatrix);
+    algorithmFPType * x = xMatrix.get();
+
+    size_t lcnt, rcnt, cnt;
+    algorithmFPType L, R, M;
+    // Use binary search to find such d that the number of verticies having distance <= d is filtNum
+    const size_t binarySearchIterNum = 20;
+    // TODO: add parallel_for
+    for (size_t i = 0; i < n; ++i)
+    {
+        L    = 0; // min possible cos distance
+        R    = 2; // max possible cos distance
+        lcnt = 0; // number of elements with cos distance <= L
+        rcnt = n; // number of elements with cos distance <= R
+        for (size_t ij = 0; ij < binarySearchIterNum; ++ij)
+        {
+            M   = (L + R) / 2;
+            cnt = 0;
+            // Calculate the number of elements in the row with value <= M
```

Comment:
Just from a high level perspective, it looks like this code will run O(binarySearchIterNum*n) to try and find the proper M which yields the right number of neighbors. binarySearchIterNum will be log(1/eps) (from here https://github.com/oneapi-src/oneDAL/pull/2875/files#r1720389061) Would a binary search tree be faster?  At some point also wouldn't just sorting the array may also be faster (when log(n) < log(1/eps))?